blocks/extensions/monitoring.py

```diff
@@ -122,22 +134,39 @@ def do(self, callback_name, *args):
         aggregation buffer and instructs the training algorithm what
         additional computations should be carried at each step by adding
         corresponding updates to it. In all other cases it writes
-        aggregated values of the monitored variables to the log.
+aggregated values of the monitored variables to the log.
 
         """
+        data, args = self.parse_args(callback_name, args)
         if callback_name == 'before_training':
             if not isinstance(self.main_loop.algorithm,
                               DifferentiableCostMinimizer):
                 raise ValueError
             self.main_loop.algorithm.add_updates(
-                self._buffer.accumulation_updates)
-            self._buffer.initialize_aggregators()
+                self._variables.accumulation_updates)
+            self.main_loop.algorithm.add_updates(
+                self._required.accumulation_updates)
```

Comment:
I don't think that the order matters here. These updates will act on different shared variables.